src/pow.rs

```diff
@@ -0,0 +1,1510 @@
+impl_try_from_signed_type_for_uint256!(i16);
+impl_try_from_signed_type_for_uint256!(i32);
+impl_try_from_signed_type_for_uint256!(i64);
+impl_try_from_signed_type_for_uint256!(i128);
+
+/// Error from `TryFrom<signed type>` implementations, occurs when input is negative.
+#[derive(Debug)]
+pub struct TryFromError(i128);
+
+impl fmt::Display for TryFromError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "attempt to create unsigned integer type from negative number: {}", self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for TryFromError {}
+
+impl Add for U256 {
+    type Output = Self;
+    fn add(self, rhs: Self) -> Self { self.wrapping_add(rhs) }
+}
+
+impl Sub for U256 {
+    type Output = Self;
+    fn sub(self, rhs: Self) -> Self { self.wrapping_sub(rhs) }
+}
+
+impl Mul for U256 {
+    type Output = Self;
+    fn mul(self, rhs: Self) -> Self { self.wrapping_mul(rhs) }
+}
+
+/// Implements mul by unsigned int in both directions. Requires `u64::from($int)`.
+macro_rules! impl_mul_for_uint256 {
+    ($int:ident) => {
+        impl Mul<$int> for U256 {
+            type Output = Self;
+            fn mul(self, rhs: $int) -> Self { self.mul_u64(u64::from(rhs)) }
+        }
+
+        impl Mul<U256> for $int {
+            type Output = U256;
+            fn mul(self, rhs: U256) -> U256 { rhs.mul_u64(u64::from(self)) }
+        }
+    };
+}
+impl_mul_for_uint256!(u8);
+impl_mul_for_uint256!(u16);
+impl_mul_for_uint256!(u32);
+impl_mul_for_uint256!(u64);
+
+impl Div for U256 {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { self.div_rem(rhs).0 }
+}
+
+impl Rem for U256 {
+    type Output = Self;
+    fn rem(self, rhs: Self) -> Self { self.div_rem(rhs).1 }
+}
+
+impl Not for U256 {
+    type Output = Self;
+
+    fn not(self) -> Self { U256(!self.0, !self.1) }
+}
+
+impl Shl<u32> for U256 {
+    type Output = Self;
+    fn shl(self, shift: u32) -> U256 { self.wrapping_shl(shift) }
+}
+
+impl Shr<u32> for U256 {
+    type Output = Self;
+    fn shr(self, shift: u32) -> U256 { self.wrapping_shr(shift) }
+}
+
+impl fmt::Display for U256 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.is_zero() {
+            f.pad_integral(true, "", "0")
+        } else {
+            self.fmt_decimal(f)
+        }
+    }
+}
+
+impl fmt::Debug for U256 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { write!(f, "{:#x}", self) }
+}
+
+macro_rules! impl_hex {
+    ($hex:ident, $fmt:literal) => {
+        impl $hex for U256 {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                if f.alternate() {
+                    write!(f, "0x")?;
+                }
+                write!(f, $fmt, self.0)?;
+                write!(f, $fmt, self.1)?;
+                Ok(())
+            }
+        }
+    };
+}
+impl_hex!(LowerHex, "{:032x}");
+impl_hex!(UpperHex, "{:032X}");
```

Comment:
Oh that was on purpose, I thought that for `Target` and `Work` we always wanted to see all the bits. I guess if reading the `U256` code alone it would seem strange but really this is just implemented on `U256` to save code duplication. Could improved by doing either
1. Have the macro take a type and implement for `Target` and `Work` directly - remove implementation for `U256` or add a 'correct' implementation
2. Just document why its implemented like this and leave it at that

Which would you prefer?

Reply:
Actually implementing it correctly is pretty straightforward. Just write the hex to temporary `[u8; 64]` buffer, convert to str using `from_utf8` and call `f.pad_integral()`. I'm preparing some hex PRs to make things easier so maybe wait for them.

Reply:
Cool, I'll leave it as is and improve along with other hex improvements to come.